src/main/java/pl/zaprogramujzycie/mma/services/MedicineService.java

```diff
@@ -0,0 +1,79 @@
+package pl.zaprogramujzycie.mma.services;
+
+import org.springframework.data.domain.Page;
+import org.springframework.data.domain.PageRequest;
+import org.springframework.data.domain.Pageable;
+import org.springframework.data.domain.Sort;
+import org.springframework.stereotype.Service;
+
+import pl.zaprogramujzycie.mma.dto.request.MedicineRequest;
+import pl.zaprogramujzycie.mma.dto.response.MedicineResponse;
+import pl.zaprogramujzycie.mma.dto.response.MedicinesResponse;
+import pl.zaprogramujzycie.mma.entity.Medicine;
+import pl.zaprogramujzycie.mma.repository.MedicineRepository;
+import pl.zaprogramujzycie.mma.utils.mappers.MedicineMapper;
+
+import java.security.Principal;
+
+@Service
+public class MedicineService {
+
+    private final MedicineRepository repository;
+
+    private MedicineMapper mapper;
+
+    public MedicineService(MedicineRepository repository) {
+        this.repository = repository;
+    }
+    //ToDo add logic of assigning owner form Family id
+    public MedicineResponse save(MedicineRequest request) {
+        MedicineResponse response = MedicineMapper.mapToMedicineDTO(repository.save(MedicineMapper.mapToMedicine(request)));
+        return response;
+    }
+
+    public MedicineResponse findById(long id, Principal principal){
+        Medicine response = findEntity(id, principal);
+        if (response != null) {
+            return MedicineMapper.mapToMedicineDTO(response);
+        } else {
+            return null;
+        }
+    }
+
+    public MedicinesResponse findAll(Principal principal, Pageable pageable){
+        Page<Medicine> medicine = repository.findByOwner(principal.getName(), PageRequest.of(
+                pageable.getPageNumber(),
+                pageable.getPageSize(),
+                pageable.getSortOr(Sort.by(Sort.Direction.ASC, "name"))));
+        Page<MedicineResponse> page = MedicineMapper.mapToResponsePage(medicine);
+        MedicinesResponse response = new MedicinesResponse(page.getContent());
+       return response;
+    }
+
+//ToDo if(PAO != 0) change expiration Date to (request timestamp + PAO)
+    public MedicineResponse partialUpdate(long id, MedicineRequest request, Principal principal) {
```

Comment:
Takie operacje powinny byc w transakcji https://www.baeldung.com/transaction-configuration-with-jpa-and-spring